1595 include docs in ui

alexandre-allard opened the pull request:
**Component**: buildchain, doc, ui

**Context**: we want to embed the documentation inside the metalk8s-ui container image

**Summary**: split build of the UI NodeJS code and container image, then embed doc inside the container image.

**Acceptance criteria**: Nothing is broken and you can access the documentation through the UI ip:port/docs


---

Closes: #1595

alexandre-allard committed:
build: ensure build root is created for image builders

Add a task dependency on build root creation
for local image builder tasks, otherwise it fails
if the root directory does not exist.

alexandre-allard committed:
build: generate docs into _build/docs dir

Instead of generating the doc in `docs/_build/`,
we now build it directly in `_build/` like all the
other components.
This change is needed to embed docs in metalk8s-ui
container

Refs: #1595

alexandre-allard committed:
build: split build of metalk8s-ui app & image

Build NodeJS stuff before generating the
metalk8s-ui image. It will allow to embed the
documentation and the NodeJS code when generating
the metalk8s-ui container image.

Refs: #1595

alexandre-allard committed:
build: build metalk8s-ui from artifacts

We now build the MetalK8s UI image from
the artifacts generated from the doc & the ui tasks
So, now we also ship the documentation inside the
metalk8s-ui container image.
The docs is accessible at the following URL:
  container_ip:container_port/docs

Refs: #1595

bert-e commented:
# Hello alexandre-allard-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Branches have diverged



This pull request's source branch `improvement/1595-include-docs-in-ui` has diverged from
`development/2.4` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.4` into `improvement/1595-include-docs-in-ui`
* Rebase `improvement/1595-include-docs-in-ui` onto `origin/development/2.4`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

slaperche-scality commented on buildchain/buildchain/builder.py:
```diff
@@ -67,6 +67,9 @@ def _builder_image(
     build_context=constants.ROOT,
     file_dep=[
         constants.ROOT/'tox.ini', constants.ROOT/'docs/requirements.txt'
+    ],
```
Good catch!

But this applies to all the builder images (we always touch a file under `_build`) so I would do the fix in the `_builder_image` helper (that way, it will applies to every image).

```diff
@@ -39,6 +39,7 @@ def _builder_image(
         dockerfile=dockerfile,
         destination=config.BUILD_ROOT,
         save_on_disk=False,
+        task_dep=['_build_root']
         **kwargs
     )
```

alexandre-allard replied:
Yeah, I saw it afterward, forgot to change it.

slaperche-scality commented on buildchain/buildchain/builder.py:
```diff
@@ -82,9 +82,24 @@ def _builder_image(
+    },
+    file_dep=[
+        constants.ROOT/'ui'/'entrypoint.sh',
+    ]
```
minor: can be squashed

```python
    build_args={'NODE_IMAGE_VERSION': versions.NODEJS_IMAGE_VERSION},
    file_dep=[constants.ROOT/'ui'/'entrypoint.sh']
```

slaperche-scality commented on buildchain/buildchain/ui.py:
```diff
@@ -0,0 +1,73 @@
+# coding: utf-8
+
+"""Tasks to build the MetalK8s UI.
+"""
```
can be squashed: `"""Tasks to build the MetalK8s UI."""`

or you could a a little schema like:

```
┌────┐    ┌───────┐    ┌───────┐
│ ui │───>│ mkdir │───>│ build │
└────┘    └───────┘    └───────┘
```
as we do for other module.

alexandre-allard replied:
Oh yeah, wanted to add the schema like the other modules, just forgot to do it before pushing. Will do

slaperche-scality commented on buildchain/buildchain/ui.py:
```diff
@@ -0,0 +1,73 @@
+
+
+def task__ui_build() -> types.TaskDict:
+    """Build the MetalK8s UI NodeJS"""
```
minor: Missing period at the end.

slaperche-scality commented on ui/Dockerfile:
```diff
@@ -1,27 +1,13 @@
 
-COPY src /home/node/src
-RUN npm run build
+COPY entrypoint.sh /entrypoint.sh
```
Why do you need to copy the entrypoint? You're mounting it when you use it to build the UI.

alexandre-allard replied:
So that we can also use that Dockerfile to build the npm stuff regardless of the buildchain.

slaperche-scality commented on buildchain/buildchain/image.py:
```diff
@@ -234,20 +234,18 @@ def _operator_image(name: str, **kwargs: Any) -> targets.OperatorImage:
     _local_image(
         name='metalk8s-ui',
-        dockerfile=constants.ROOT/'ui'/'Dockerfile',
+        build_context=config.BUILD_ROOT,
```
commit message: `s/We know build/We now build/`

slaperche-scality commented on images/metalk8s-ui/conf/nginx.conf:
```diff
@@ -0,0 +1,44 @@
+server {
```
This is a duplicate of `ui/conf/nginx.conf`, no?
Do we need to keep these two copies,

alexandre-allard replied:
Hum no, forgot to remove it.

slaperche-scality commented on buildchain/buildchain/ui.py:
```diff
@@ -70,4 +73,22 @@ def clean() -> None:
+        'targets': [target],
+        'task_dep': ['_build_root'],
+        'file_dep': [source],
+        'clean': [clean(target)],
```
want to delete `target`, then you can simply use `'clean': True`.

> A common scenario is a task that needs to “revert” its actions. A task may include a clean attribute. This attribute can be True to remove all of its target files.

See https://pydoit.org/cmd_other.html#clean

You only need a custom callable when you want to delete stuff that aren't in `targets`.

alexandre-allard replied:
Nice. Will do.

slaperche-scality commented on buildchain/buildchain/ui.py:
```diff
@@ -36,7 +39,7 @@ def task__ui_mkdir_build_root() -> types.TaskDict:
 
 def task__ui_build() -> types.TaskDict:
-    """Build the MetalK8s UI NodeJS"""
+    """Build the MetalK8s UI NodeJS code"""
```
minor: missing period.

NicolasT commented on images/metalk8s-ui/Dockerfile:
```diff
@@ -0,0 +1,12 @@
+RUN rm -rf /usr/share/nginx/html/*
+
+COPY ui /usr/share/nginx/html/
+COPY docs/html /usr/share/nginx/docs
```
No, not fond of this. `docs` are part of *DocumentRoot* as is `ui`. So, copy `ui` into `/usr/share/nginx/html`, copy `docs` into `/usr/share/nginx/html`, set up `/usr/share/nginx/html` as the server root, and that should basically be it.

alexandre-allard replied:
Ok, wasn't sure of how we wanted to do this, will do.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/1595-include-docs-in-ui` into
`development/2.4`
* [w/2.5/improvement/1595-include-docs-in-ui](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/improvement/1595-include-docs-in-ui) will be merged into `development/2.5` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

* @slaperche-scality

slaperche-scality commented on buildchain/buildchain/docs.py:
```diff
@@ -45,28 +45,35 @@
 DocTarget = namedtuple('DocTarget', ('name', 'command', 'target'))
-DOC_BUILD_DIR : Path = constants.ROOT/'docs/_build'
-DOC_PDF_FILE  : Path = DOC_BUILD_DIR/'latex/{}.pdf'.format(config.PROJECT_NAME)
+DOC_PDF_FILE : Path = constants.DOCS_BUILD_ROOT/'latex/{}.pdf'.format(
```
minor: when the opening parenthesis is on its own line, the closing as well:

```python
DOC_PDF_FILE : Path = constants.DOCS_BUILD_ROOT/'latex/{}.pdf'.format(
    config.PROJECT_NAME
)
```

slaperche-scality commented on buildchain/buildchain/ui.py:
```diff
@@ -0,0 +1,73 @@
+
+    return {
+        'actions': [build_ui],
+        'title': utils.title_with_target1('UI BUILD'),
```
s/UI BUILD/NPM BUILD/ maybe

slaperche-scality commented on buildchain/buildchain/ui.py:
```diff
@@ -4,15 +4,16 @@
-│ mkdir │───>│ build ui │
-└───────┘    └──────────┘
+┌────┐    ┌───────┐    ┌──────────┐    ┌───────────┐
+│ ui │───>│ mkdir │───>│ ui build │───>│ ui config │
```
UI configuration (that way you don't need to add `config` to the dict`)

slaperche-scality commented on buildchain/buildchain/ui.py:
```diff
@@ -0,0 +1,73 @@
+Overview:
+
+┌───────┐    ┌──────────┐
+│ mkdir │───>│ build ui │
```
build UI

slaperche-scality reviewed (commented):
only a few minor remarks :wink:

alexandre-allard committed:
build: handle umask for mkdir target

Add the possibility to change the umask when
creating a directory.
This is needed to build the UI as we do it
with a specific user `node`, which needs to
be able to write in the volume which is created
by the local user. So we need to enforce the mode
to 777 on the directory.

Refs: #1595

bert-e commented:
# History mismatch



Merge commit #7641199595f36c507f1bd77e3f41f3f086936ed9 on the integration branch
`w/2.5/improvement/1595-include-docs-in-ui` is merging a branch which is neither the current
branch `improvement/1595-include-docs-in-ui` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `improvement/1595-include-docs-in-ui` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

alexandre-allard commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/1595-include-docs-in-ui` into
`development/2.4`
* [w/2.5/improvement/1595-include-docs-in-ui](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/improvement/1595-include-docs-in-ui) will be merged into `development/2.5` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

* @slaperche-scality

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

alexandre-allard commented:
/approve

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/44bd22461800b611f1572c4bda82f269f82102da) did not succeed in branch w/2.5/improvement/1595-include-docs-in-ui.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye alexandre-allard-scality.
